Add burn method

markusait opened the pull request:
I added a burn method and the following tests:

- Alice deploys the module, proves ownership and mints tokens to her account
- Alice burns tokens, and it is checked that the correct amount matches the balance 

You can find a cli implementation for manual tests here: https://github.com/stablyio/USDS_Libra

Still todo:

- decide whether capability should be passed to revert burns if a user is blacklisted for example
- Test that burning more that Alice balance should fail ( I could not get the `6_blacklisted_user_transfer_token_should_fail.mvir` script to work
- Test that burning negative amount is not possible
- Test that burning from Burner that has not published yet fails

Looking forward to getting your feedback and make changes accordingly.

markusait committed:
added first version of burn method

markusait commented on testscripts/9_regular_user_can_burn_his_tokens.mvir:
```diff
@@ -0,0 +1,378 @@
+        // Making sure more than 0 is burned
+        assert(move(burned_value) != 0, 3);
+        return;
+    }
```
Burning by resource here not sure if the != 0 assert is necessary. It is less code than the other burn method but I am not sure what is better in regards to security and performance. 
I have added a basic test for this function on line 363

peteremiljensen replied:
The assertion shouldn't be necessary. I like this version better as it is more concise. It also seems more secure as the type system now guarantees that you can only burn what you own.

peteremiljensen commented:
Looks good. However, somebody should include it in the main source and in the other tests before I can approve it. I might be able to do it tomorrow or at the beginning of next week

markusait commented:
Hi, I didn't get to work on this on the weekend but I just updated the PR and added the burn method to all tests and the main source. I could also add the burn test to the other test files. Is there anything else that needs to be done?
Edit: It seems like my VSCode added a bunch of linting changes which makes the PR hard to read. Is there a lint command I can run to revert these changes?

peteremiljensen commented:
Thank you very much Markus. We’re almost there. Can you just make sure that your own burn test includes the newest eToken? Currently, it is missing the latest changes such as revocation. Again thank you very much

Peter

markusait commented:
Ok, I incorporated the latest changes, rebased and ran the tests. I saw that you can hide the space changes so that my PR is actually readable now. 
It's been my pleasure to contribute really cool stuff you guys are building, looking forward to where this project is going.

peteremiljensen commented:
Thank you very much for your contribution and time put into it 👍
